ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java

```diff
@@ -306,4 +323,146 @@ public int decompress(byte[] compressedBuffer, int compressedBufferOffset, int c
+      compressedBuffer.reset();
+    }
+  }
+
+  /**
+   * Compress the buffer specified in {@code sourceBuffer}.  The entire sourceBuffer will be read,
+   * so set the position and limit to the range of data to read in sourceBuffer before calling this method.
+   * After calling, sourceBuffer position will be advanced to the buffer limit, and the compressedBuffer position
+   * will be advanced to the end of the compressed binary.
+   * If compression failed, the sourceBuffer position will not be advanced.
+   *
+   * @param sourceBuffer The source/uncompressed data to compress.  It cannot be null or empty.
+   * @param compressedBuffer The compressed buffer where the compressed data will be written to.  Its size must be
+   *                       at least the size return from getCompressBufferSize() or it may fail due to buffer size.
+   * @return The actual compressed data size in bytes.
+   * @throws CompressionException Throws this exception if compression has internal failure.
+   */
+  @Override
+  public int compress(ByteBuffer sourceBuffer, ByteBuffer compressedBuffer) throws CompressionException {
+    Utils.checkNotNullOrEmpty(sourceBuffer, "sourceBuffer cannot be null or empty.");
+    Utils.checkNotNullOrEmpty(compressedBuffer, "compressedBuffer cannot be null or empty.");
+
+    int overheadSize = VERSION_AND_ORIGINAL_SIZE_SIZE + ALGORITHM_NAME_LENGTH_SIZE + getAlgorithmNameBinary().length;
+    if (compressedBuffer.remaining() < overheadSize) {
+      throw new IllegalArgumentException("compressedBuffer " + compressedBuffer.remaining() + " is too small.");
+    }
+
+    // Write version, algorithm name, and source data size.
+    int compressedBufferStartPosition = compressedBuffer.position();
+    int sourceBufferStartPosition = sourceBuffer.position();
+    try {
+      byte[] algorithmNameBinary = getAlgorithmNameBinary();
+      compressedBuffer.put((byte) 1);
+      compressedBuffer.put((byte) algorithmNameBinary.length);
+      compressedBuffer.put(algorithmNameBinary);
+      compressedBuffer.putInt(sourceBuffer.remaining());
+
+      // Apply compression and store the output in the compressed buffer.
+      // Note: compressNative() uses less memory than buffer size and that's why it returns the actual compressed size.
+      int sourceBufferLimit = sourceBuffer.limit();
+      int compressedDataLength =
+          compressNative(sourceBuffer, sourceBuffer.position(), sourceBuffer.remaining(), compressedBuffer,
+              compressedBuffer.position(), compressedBuffer.capacity() - compressedBuffer.position());
+
+      // compressNative() may or may not alter the position indexes, so set them just in case.
+      sourceBuffer.position(sourceBufferLimit);  // Advance the sourceBuffer to limit to indicate entire buffer read.
+      compressedBuffer.position(compressedBufferStartPosition + overheadSize + compressedDataLength);
+      return (overheadSize + compressedDataLength);
+    } catch (Exception ex) {
```

Comment:
any other exception rather than CompressionException?

Reply:
Yes, to comply with the documentation "if compress() failed, the indexes will not change".  The indexes advances only if succeeded.  The goal is to avoid leaving bytebuffer in an unpredictable position when exception is thrown.